src/linux_mcp_server/tools/system_info.py

```diff
@@ -1,138 +1,290 @@
+def _apply_list_filter(
+    items: list[t.Any],
+    filter_key: str,
+    filter_values: list[str] | None,
+) -> list[t.Any]:
+    """
+    Helper to filter a list of dictionaries by a specific key.
+
+    If filter_values is None, return all items.
+    """
+    if filter_values is None:
+        return items
+
+    return [item for item in items if isinstance(item, dict) and item.get(filter_key) in filter_values]
+
+
+# Pydantic Models for structured output
+class SystemInfo(BaseModel):
+    """System information model."""
+
+    hostname: str | None = None
+    operating_system: str | None = None
+    os_version: str | None = None
+    kernel_version: str | None = None
+    architecture: str | None = None
+    uptime: str | None = None
+    boot_time: str | None = None
+
+
+class LoadAverage(BaseModel):
+    """Load average information."""
+
+    one_min: float = 0.0
+    five_min: float = 0.0
+    fifteen_min: float = 0.0
+
+
+class CPUInfo(BaseModel):
+    """CPU information model."""
+
+    cpu_model: str | None = None
+    logical_cores: int | None = None
+    physical_sockets: int | None = None
+    physical_cores_per_processor: int | None = None
+    frequency_mhz: float | None = None
+    load_average: LoadAverage | None = None
+    cpu_usage: str | None = None
+
+
+class MemoryStats(BaseModel):
+    """Memory statistics model."""
+
+    total: int = 0
+    used: int = 0
+    free: int = 0
+    shared: int = 0
+    percent: float = 0.0
+    available: int | None = None
+    buffers: int | None = None
+    cached: int | None = None
+
+
+class MemoryInfo(BaseModel):
+    """Memory information model."""
+
+    ram: MemoryStats
+    swap: MemoryStats
+
+
+class DiskPartition(BaseModel):
+    """Disk partition information."""
+
+    device: str
+    mountpoint: str
+    size: int
+    used: int
+    free: int
+    percent: float
+    filesystem: str | None = None
+
+
+class DiskIOStats(BaseModel):
+    """Disk I/O statistics."""
+
+    read_bytes: int = 0
+    write_bytes: int = 0
+    read_count: int = 0
+    write_count: int = 0
+
+
+class DiskUsage(BaseModel):
+    """Disk usage information model."""
+
+    partitions: list[DiskPartition]
+    io_stats: DiskIOStats
+
+
+class PCIDevice(BaseModel):
+    """PCI device information."""
+
+    address: str
+    vendor_id: str | None = None
+    device_id: str | None = None
+    class_id: str | None = None
+    description: str | None = None
+
+
+class USBDevice(BaseModel):
+    """USB device information."""
+
+    bus: str
+    device: str
+    vendor_id: str | None = None
+    product_id: str | None = None
+    description: str | None = None
+    manufacturer: str | None = None
+
+
+class DeviceInfo(BaseModel):
+    """Device information model for PCI and USB devices."""
+
+    pci_devices: list[PCIDevice] = Field(default_factory=list)
+    pci_device_count: int = 0
+    usb_devices: list[USBDevice] = Field(default_factory=list)
+    usb_device_count: int = 0
```

Comment:
The count parameters should be properties, not parameters.
```suggestion
    pci_devices: list[PCIDevice] = Field(default_factory=list)
    usb_devices: list[USBDevice] = Field(default_factory=list)
    
    @property
    def pci_device_count(self) -> int:
        return len(self.pci_devices)
        
    @property
    def usb_device_count(self) -> int:
        return len(self.usb_devices)
```